### This folder contains the code needed to go from raw illumina data to called SNPs

- The scripts 'demultiplex.plate1.sh' and 'demultiplex.plate2.sh' contain the code used to demultiplex all of the individual samples pooled on this Illumina NextSeq2000 run. These two scripts were run separately, because the 155 samples on this run were dual-indexed, meaning they had a unique combination of plate index and barcode (i.e., barcodes overlapped between plates 1 and 2). So, if you were to demultiplex all of them together at the same time, samples would be mixed. Instead, samples were first demultiplexed by plate (based on plate index) and then each plate was further demultiplexed into individual fastq files according to 'plate1.barcodes.txt' and 'plate2.barcodes.txt' for each respective plate.

- The script 'run.stacks.sh' then executes a run of the program [Stacks](https://catchenlab.life.illinois.edu/stacks/), which involves mapping each sample against the publicly available *Zosterops japnicus* [reference genome](https://www.ncbi.nlm.nih.gov/datasets/genome/GCA_017612475.1/), collating loci, and calling SNPs. I then output a completely unfiltered SNP dataset which was used as input for a downstream filtering pipeline in R found [here](https://devonderaad.github.io/zosterops.rad/zost.radseq.filtering.html).
